### PR TITLE
Pspin dgraphs

### PR DESCRIPTION
**Pull request by smcantab**
This pull request is for the implementation/discussion of disconnectivity graphs calculations for the p-spin spherical model.

- [x] check that database i/o works
- [x] figure out whether zero mode is problematic during transition state search (if so implement orthogonalize function in system class)
- [ ] figure out why rms O(1e2) when running NEB
- [x] write scripts for production run (sample minima and connect all)
- [ ] add drawing feature to gui (using a simpler heat map)
- [x] write general function to compute interaction matrix for p not 3
- [x] consider using jit (numba) to speed up slowest calls in connect all

### CODE REVIEW

**Comment by smcantab** [comment]
In `get_orthogonalize_to_zero_eigenvectors` I return a function that orthogonalizes the vector with respect to the coordinate vector (or in other words the radial component). I am not sure whether this orthogonalizes wrt the zero mode eigenvector but certainly whichever eigenvector we follow must be also orthogonal to the radial direction. This seems to stabilise TS refinement when using a lower neb image density (that if I remember correctly was blowing things up).

**Comment by js850** [comment]
That's the way I would do it :+1:

**Comment by smcantab** [comment]
I'm thinking of rewriting the parallel implementation of double ended
connect and bh using mpi4py, do you guys think it's worth spending a day or
two on it?

**Comment by js850** [comment]
This example shows the current way of doing things in parallel https://github.com/pele-python/pele/tree/master/examples/parallel_pele

It has some drawbacks as you know.  mpi4pi would be easier to run on the cluster, but the hard part is that sqlite doesn't allow concurrent modifications.  doing it properly would be involved.  i wouldnt spend the time unless you really need to

**Comment by smcantab** [comment]
ok thanks. I'll use the pyro implementation and I'll rewrite it only if I
get to the point that pyro can't cope anymore.

Regarding the concurrent access I think that it would be ok to do. One way
is to use blocking collective communication, so everyone, including master
does his own thing, then master sorts out all incoming communications and
waits until they're all done. This should be massively scalable but not
very efficient.
The other option is master constantly listening and using point-to-point
non-blocking communication (Isend, Irecv etc in mpi4py), instead of
scatter/broadcast,/gatherall but it would be a bit ugly, I agree. The
latter would do exactly what Pyro is doing but possibly more scalable.

**Comment by smcantab** [comment]
Systems for which p is even have an additional symmetry, therefore I need to fix one spin to not get double of everything (s*s = -s*-s). I have made a number of changes that involve the frozen potential cpp files and the pspin cpp potential. 

I don't think this is working correctly yet. When I open the gui and look at the normal modes I don't have a zero mode but instead always a small negative mode, but the energy clearly shows that it is a minimum which means that the mode should be positive. It must have to do with the fact that the normal modes are computed for the reduced matrix and the energy comes from the full coordinates. But I don't know whether to worry or not, I think I should worry.

The changes that I had to do are the following.

1. I had to make sure that frozen potential wrapper uses a reference of the coordinates, not a copy (I implement this artificially with an additional copy). This is necessary because the normalization step happens in the potential. This only involves an additional memcpy, if the coordinates are not modified in the potential then nothing changes. (I also replace loops with std::copy where possible, this should improve performance a little in FrozenPotentialWrapper)
2. I had to add a flag to pspin to know where spin 0 has been frozen. This is necessary because when normalizing I need to make sure that the value of spin 0 remains unaltered. Therefore I had to implement the normalize spin function in a slightly more complicated way.
3. I had to modify the orthogonalize function and I think this is where the problem is. I am not sure I have implemented it correctly.

Any suggestions would be really helpful.

**Comment by smcantab** [comment]
this is what I mean

snapshot1

**Comment by js850** [comment]
is freezing a spin really the correct solution there?  It works for discrete spins (s = +-1), but will it also work for continuous spins?  Is't the symmetry discrete (s => -s)?

**Comment by js850** [comment]
Maybe you could just account for this symmetry in mindist and exact_match

**Comment by smcantab** [comment]
the symmetry is such that for each array x with energy E, there is an array
-x also with energy E. The result is that when I look for the minima I find
all pairs and when I connect them I get disconnectivity graph where the LHS
and RHS of the tree are mirror images.

By using exact_match to treat x and -x as the same minimum wouldn't I incur
in bigger issues when running connect all? For instance I can't think of
any reason why the shortest path from x1 to x2 could not be x1 -> -x1 ->
x2, so I would find two transition states and add both of them but only add
one minimum out of x1 and -x1 to the database.

**Comment by js850** [comment]
That's what mindist is for.  Given two arrays x1 and x2, you transform x2 to be as close as possible to x1.

```
d1 = dist(x1, x2)
d2 = dist(x1, -x2)
if d1 < d2:
    return d1, x1, x2
else:
    return d2, x1, -x2
```

I really don't think freezing one spin is correct

**Comment by smcantab** [comment]
I also agree that fixing the spin is not the right thing to do. Fixing it
means removing a whole continuous degree of freedom. Instead we just want
to restrict ourself to 1/2 of the surface of the hypesphere. The question
is how to do that.

**Comment by js850** [comment]
LJ clusters have the same symmetry.  Mirror symmetry x => -x gives the same energy

**Comment by smcantab** [comment]
you're right, that's the way to go :)

**Comment by smcantab** [comment]
ok I have reversed the unnecessary changes, although I have retained some aesthetic ones and those that are likely to improve performance. It runs like a charm now, p=4 landscapes seem to be a lot more interesting from what I have seen so far.

**Comment by js850** [comment]
Great!

**Comment by smcantab** [comment]
Happy new year! I have now written scripts to run connect_all and basin_hopping in parallel using Pyro4

File: playground/pspin/start_connect_worker.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    #get interactions from database
+    dbname = "pspin_spherical_p{}_N{}.sqlite".format(p,nspins)
+    db = Database(dbname, createdb=False)
+    interactions = db.get_property("interactions").value()
+    db_nspins = db.get_property("nspins").value()
+    db_p = db.get_property("p").value()
+    assert db_nspins == nspins
+    assert db_p == p
+    #close this SQLAlchemy session
+    db.session.close()
```
Comment (js850) [inline playground/pspin/start_connect_worker.py]: It's good that you close the db to avoid running into problems having multiple instances open.  To be even more safe you might want to put this section in a separate function so `db` can be garbage collected.

Comment (smcantab) [inline playground/pspin/start_connect_worker.py]: good point!

**Comment by smcantab** [comment]
OK, the state of affairs is the following: I have written and tested the parallel basin hopping and connect all on my machine and the cluster BUT I realised that for p=3 the scaling of the operations for a full quench is something like N^4.3. Cubic complexity comes from the (N choose 3) and the remaining linear part must be due to the length of the quench (that's my best guess).

Since we need at least N=200 to get the glassy landscape and all the fancy properties that are predicted by the theory, this cpu implementation is hopeless. The guys at NYU used gpus for their previous work, and now I see why.

Henceforth, I am thinking of using the "thrust" library from the cuda toolkit, it has a [`reduce_transform`](http://docs.thrust.googlecode.com/hg/group__transformed__reductions.html#ga0d4232a9685675f488c3cc847111e48d) function (basically map-reduce) that should work out of the box without going much into the gpu-coding realm.

Opinions/suggestions? The cool thing about thrust is that the same code can run on both GPUs (cuda) and CPUs (openMP), depending on how you compile it.

**Comment by js850** [comment]
Before spending a lot of time implementing it yourself I would spend at least a while trying to get someone else's GPU implementation.

**Comment by smcantab** [comment]
As far as I can tell there are no public implementations of this particular potential (besides the one we can get from NYU, but that one is specific to p=3 and might be hard to fit within pele if it's a low-level implementation, dunno). I have spent some time today looking through `thrust` quick-start guide and documentation and examples, I have also implemented minimal tests on my computer for the `transform_reduce` function to get a better feel of how things work (although came to the conclusion that this is not the function that'll do it). I also looked a little bit around to see what's out there for the integration of cuda code in python through cython, my first impression is the following:

1. thrust should make the implementation of the potential possible without need of any cuda specific syntax. Obviously the fact that it will run on gpus does not mean that it will be fast but let's worry later about that. Now we must worry about 2 things
    * while we can keep the array of interactions stored on the device (gpu), we must transfer the list of spin from the host to the device at each iteration because the minimizers are not implemented for gpus. This could be slow, but nspins<1000 so it's probably not a big deal.
    * how to get the specific combination indexes without storing them, they would take p*(N choose p)*sizeof(int). The device memory is limited and we must already store at least the (N choose p) interactions (although at the moment we store the N^p array, so while easy to do we'd need to figure out that one too)
2. we must figure out how to actually manage the compilation within pele and how to organize the code. It might be nice to have the structure to support cuda potentials in general though. Hoomd-blue does it and I also found another (less confusing) minimal wrapping [example](https://github.com/rmcgibbo/npcuda-example/tree/master/cython) 

In any case it's gonna be some work...and it'll definitely going to need another pull request :sweat_smile:

**Comment by js850** [comment]
It sounds like a lot of work.  Think carefully about whether you want to invest that time.  IT might be worth it for the experience though.  b.t.w. Rosemary implemented a cuda version of LBFGS for GMIN which you can make use of.  Talk to her if you have questions

**Comment by smcantab** [comment]
ok, so I have emailed Rosy about cuda lbfgs (you're cc'd). Definitely worth
for the experience and would definitely be a slick addition to pele. The
cuda lbfgs would be really good to have when playing with neural networks.
About that see below...

I spoke to the NYU people and they used the (cu)torch lua interface to
implement the pspin potential. Here comes the other important point:
ultimately the plan is to interface torch with pele to make dgraphs on
neural networks. This is the core of this project I guess.

I am going to open two issues/PR to do these 2 things: cuda support for
pele and pele-torch interface (this should probably be a new library, which
we can call "Porch" the pele-torch interface 😂)

**Comment by smcantab** [comment]
actually torch has a library called optim that implements lbfgs and other minimizers, so maybe the least painful approach is to just interface our python layer with the lua minimizer+potential, completely bypassing the our cpp backend. This would address both points above

**Comment by smcantab** [comment]
sorry for the spam today, I had a long chat with Julian and looked around for options. We came to the conclusion that TensorFlow might the best option and we can use it to write an implementation of the pspin model that runs on gpus that would fit within a BasePotential and that would ultimately allow us to code some ML stuff.

**Comment by js850** [comment]
two birds, one stone, I like it :+1: .  TensorFlow is probably a good idea given that it has easy access to python.  Torch is lua based (ugh), but you could probably get help from Andy.  Maybe Andy could give a suggestion of one or the other.

**Comment by smcantab** [comment]
A very first implementation of the potential is available here

https://github.com/smcantab/pele/blob/pspin_tensorflow/playground/pspin/pspin_potential_tf.py

it does what it's supposed to, but I still don't understand some things so it'll probably change quite a bit, especially with respect to the way the gradient will be computed: ideally I would like to use the built in compute_gradient function but I don't understand how to do that yet (I didn't find many examples). I'll have to look around a bit more.

**Comment by smcantab** [comment]
I think this is good to be merged, I have put the system class in systems and moved the parallel scripts to the examples under examples/parallel_pele/pspin. Let me know if there is any objection

**Comment by js850** [comment]
hmm.  Travis is failiing.  but it doesn't look to be connected to this PR

Also, it would be great (but not necessary) to create a little example script in the examples folder.  I see the parallel one, but maybe you could add a serial one also

**Comment by smcantab** [comment]
ok, I have a serial one in learnscapes, I'll adapt it. I am not sure what
that error is about either. I'm struggling with travis today, it's giving
me a hard time with nested_sampling as well :) I'm trying to set it up
there too
